{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "U3aG1MwnuXaW",
        "SKyJNQEXrgE7"
      ],
      "authorship_tag": "ABX9TyPIONulkbrxTv0TMlNUtrTA",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/rcs-science-research/Dendritic-Branching-GSEA/blob/main/Dendritic_Branching_Associated_Pathways_GSEA.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Identifying Gene Sets Associated with Dendritic Branching using Gene Set Enrichment Analysis (GSEA)"
      ],
      "metadata": {
        "id": "NORmwM-A44i0"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Neuronal function and connectivity are heavily influenced by the structure and complexity of dendrites, branching extensions that integrate signals from other neurons. Dendritic arborization, the process by which dendrites form and develop, is crucial for the proper function of individual neurons and the brain (McAllister, 2005). Local and cell-wide dendrite patterning are regulated by neuronal activity, with calcium influx mediated by ion channels like NMDA (N-methyl-D-aspartate) receptors and VSCCs (voltage-sensitive calcium channels) (Wong & Ghosh, 2002). Calcium signaling has been shown to regulate local dendrite structure, affecting branch dynamics and stability (Konur & Ghosh, 2005), and activate intracellular signaling proteins, such as CaMKs (calmodulin-dependent protein kinases), which modulate gene expression (Redmond et al., 2002). Still, the molecular mechanisms and underlying dendritic arborization remain to be fully understood, and are critical for understanding neural circuit development and synaptic connectivity, as well as various neurological pathologies.\n",
        "\n",
        "In this notebook, we will examine the regulation of dendritic morphology through analysis of the transcriptomic and morphological cortical mouse neuron data provided by the Allen Institute (Gouwens et al., 2020). Rather than focusing on the most correlated individual genes, which may not capture the broader effects of biological processes like signaling or metabolic pathways, we adopt a gene set-level approach to capture the coordinated activity of many genes. We will use Gene Set Enrichment Analysis, a statistical method that identifies gene sets correlated with a target phenotype. These gene sets are sets of related genes compiled in various databases that can represent biological processes like pathways or other types of functional relationships. The algorithm involves a number of steps, but the main principle is the use a weighted Kolmogorov-Smirnov-like statistic to calculate an enrichment score (ES) reflecting how correlated a set of genes is with the phenotype overall. More details about this algorithm can be found in the original paper: https://www.pnas.org/doi/10.1073/pnas.0506580102. By applying GSEA to examine gene expression and dendrite morphology, we will identify gene sets linked to calcium signaling, which regulates dendritic arborization through CaMKs and various signaling pathways that mediate dendritic growth.\n",
        "\n"
      ],
      "metadata": {
        "id": "LsVrWtz-5Q-w"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "gLQSTVOctUkb",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5776e742-0151-43ce-e911-3699073068d9",
        "collapsed": true
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n",
            "Collecting gseapy\n",
            "  Downloading gseapy-1.1.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (11 kB)\n",
            "Requirement already satisfied: numpy>=1.13.0 in /usr/local/lib/python3.11/dist-packages (from gseapy) (1.26.4)\n",
            "Requirement already satisfied: scipy in /usr/local/lib/python3.11/dist-packages (from gseapy) (1.13.1)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.11/dist-packages (from gseapy) (2.2.2)\n",
            "Requirement already satisfied: matplotlib>=2.2 in /usr/local/lib/python3.11/dist-packages (from gseapy) (3.10.0)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.11/dist-packages (from gseapy) (2.32.3)\n",
            "Requirement already satisfied: contourpy>=1.0.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (1.3.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (0.12.1)\n",
            "Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (4.55.8)\n",
            "Requirement already satisfied: kiwisolver>=1.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (1.4.8)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (24.2)\n",
            "Requirement already satisfied: pillow>=8 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (11.1.0)\n",
            "Requirement already satisfied: pyparsing>=2.3.1 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (3.2.1)\n",
            "Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.11/dist-packages (from matplotlib>=2.2->gseapy) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas->gseapy) (2025.1)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas->gseapy) (2025.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests->gseapy) (3.4.1)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests->gseapy) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests->gseapy) (2.3.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests->gseapy) (2025.1.31)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.7->matplotlib>=2.2->gseapy) (1.17.0)\n",
            "Downloading gseapy-1.1.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (549 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m549.2/549.2 kB\u001b[0m \u001b[31m7.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: gseapy\n",
            "Successfully installed gseapy-1.1.5\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "\n",
        "#Data analysis and manipulation\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "#GSEA\n",
        "!pip install gseapy\n",
        "import gseapy as gp\n",
        "from gseapy.plot import gseaplot\n",
        "from gseapy.plot import enrichment_map\n",
        "\n",
        "#Saving and loading results\n",
        "import pickle\n",
        "\n",
        "#Plotting and graphs\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "\n",
        "#Leading-edge subset analysis\n",
        "from sklearn.feature_extraction.text import CountVectorizer\n",
        "from scipy.spatial.distance import pdist, squareform\n",
        "from scipy.cluster.hierarchy import linkage, dendrogram, cut_tree, fcluster\n",
        "from sklearn.metrics import silhouette_score\n",
        "from sklearn.preprocessing import binarize"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Data Preprocessing"
      ],
      "metadata": {
        "id": "3iGHR_3_uNO1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Let's begin by loading the transcriptomic/morphological data and neuron metadata as pandas dataframes. By printing the columms, we can see the types of data in the datasets. In this analysis, we will be focusing on the gene expression data and dendrite.num_branches.\n"
      ],
      "metadata": {
        "id": "gGbMipXxaQw5"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cell_data = pd.read_hdf(\"/content/drive/Shareddrives/Lisman Laboratory/Lisman 2024/Neuro/Data/processed data/genes_morph.hdf5.lz4\")\n",
        "cell_metadata = pd.read_hdf(\"/content/drive/Shareddrives/Lisman Laboratory/Lisman 2024/Neuro/Data/processed data/metadata.hdf5.lz4\")\n",
        "print(cell_data.columns, \"\\n\", cell_metadata.columns)"
      ],
      "metadata": {
        "id": "LSptcfM2uXDR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "539b9e1e-eacb-4783-a007-adb810ddd31b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Index(['0610005C13Rik', '0610006L08Rik', '0610007P14Rik', '0610009B22Rik',\n",
            "       '0610009E02Rik', '0610009L18Rik', '0610009O20Rik', '0610010B08Rik',\n",
            "       '0610010F05Rik', '0610010K14Rik',\n",
            "       ...\n",
            "       'all_neurites.num_tips', 'basal_dendrite.num_tips',\n",
            "       'dendrite.num_nodes', 'axon.num_nodes', 'all_neurites.num_nodes',\n",
            "       'basal_dendrite.num_nodes', 'dendrite.num_branches',\n",
            "       'axon.num_branches', 'all_neurites.num_branches',\n",
            "       'basal_dendrite.num_branches'],\n",
            "      dtype='object', length=45792) \n",
            " Index(['project', 'cell_specimen_name', 'hemisphere', 'structure', 'donor_id',\n",
            "       'donor_name', 'biological_sex', 'age', 'ethnicity',\n",
            "       'medical_conditions', 'full_genotype', 'dendrite_type',\n",
            "       'apical_dendrite_status', 'neuron_reconstruction_type',\n",
            "       'cell_soma_normalized_depth', 'depth_from_pia_um', 'ephys_session_id',\n",
            "       'transcriptomics_sample_id', 'transcriptomics_batch',\n",
            "       'T-type Accession ID', 'T-type Label', 'MET-type Label'],\n",
            "      dtype='object')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Exploratory Analysis"
      ],
      "metadata": {
        "id": "y1OYXGfnsf0Y"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "We can get an overview and summary statistics of the data using `df.describe()`"
      ],
      "metadata": {
        "id": "3L1HkTh_ficB"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cell_data[\"dendrite.num_branches\"].describe()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 335
        },
        "id": "4y0IaR1qGDRm",
        "outputId": "c9b4abd5-24f8-459a-8075-c1f9e91c8d01"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "count    573.000000\n",
              "mean      55.265271\n",
              "std       22.805949\n",
              "min       16.000000\n",
              "25%       39.000000\n",
              "50%       50.000000\n",
              "75%       68.000000\n",
              "max      148.000000\n",
              "Name: dendrite.num_branches, dtype: float64"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>dendrite.num_branches</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>count</th>\n",
              "      <td>573.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>mean</th>\n",
              "      <td>55.265271</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>std</th>\n",
              "      <td>22.805949</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>min</th>\n",
              "      <td>16.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25%</th>\n",
              "      <td>39.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50%</th>\n",
              "      <td>50.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>75%</th>\n",
              "      <td>68.000000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>max</th>\n",
              "      <td>148.000000</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div><br><label><b>dtype:</b> float64</label>"
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "A histogram shows the distribution of the dendrite.num_branches:"
      ],
      "metadata": {
        "id": "OFr57_BHgEp8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sns.histplot(cell_data[\"dendrite.num_branches\"])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 467
        },
        "id": "zSXECDl6K9Ps",
        "outputId": "6e449269-72b1-406e-c158-667de63a3dd9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<Axes: xlabel='dendrite.num_branches', ylabel='Count'>"
            ]
          },
          "metadata": {},
          "execution_count": 4
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 640x480 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjsAAAGxCAYAAACEFXd4AAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAMWdJREFUeJzt3XtUVXX+//HXURBQBETiAClCRkKmaWJK1nSRURu7mM7F0rKyrEa8NmpO4a2LtzTTTCdn0uab5tQqm3ImG8XUnBAVMzNRMy84yiVUOF4QUT6/P/p5VsdbiOdwDpvnY629luez9/ns9/7U0tfa57P3x2aMMQIAALCoOt4uAAAAwJMIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNL8vF2AL6ioqNDBgwfVsGFD2Ww2b5cDAAAqwRijo0ePKiYmRnXqXPz+DWFH0sGDB9W0aVNvlwEAAKpg//79atKkyUX3E3YkNWzYUNJPgxUSEuLlagAAQGU4HA41bdrU+e/4xXg17KxZs0ZTp05Vdna28vLytGTJEvXo0cO53xijsWPHat68eSouLlanTp00Z84cJSQkOI85fPiwBg0apE8//VR16tRRr1699Prrrys4OLjSdZz96SokJISwAwBADfNLU1C8OkH5+PHjuvHGGzV79uwL7p8yZYpmzpypuXPnKisrSw0aNFDXrl118uRJ5zF9+vTRd999p+XLl2vp0qVas2aNBgwYUF2XAAAAfJzNV1Y9t9lsLnd2jDGKiYnRs88+qz/96U+SpJKSEtntdi1YsEC9e/dWTk6Orr/+em3YsEHJycmSpGXLluk3v/mN/ve//ykmJqZS53Y4HAoNDVVJSQl3dgAAqCEq+++3zz56vmfPHuXn5ys1NdXZFhoaqg4dOigzM1OSlJmZqbCwMGfQkaTU1FTVqVNHWVlZ1V4zAADwPT47QTk/P1+SZLfbXdrtdrtzX35+viIjI132+/n5KTw83HnMhZSVlamsrMz52eFwuKtsAADgY3z2zo4nTZw4UaGhoc6Nx84BALAunw07UVFRkqSCggKX9oKCAue+qKgoFRYWuuw/ffq0Dh8+7DzmQkaPHq2SkhLntn//fjdXDwAAfIXPhp34+HhFRUUpIyPD2eZwOJSVlaWUlBRJUkpKioqLi5Wdne08ZuXKlaqoqFCHDh0u2ndAQIDzMXMeNwcAwNq8Omfn2LFj2rVrl/Pznj17tHnzZoWHhys2NlZDhw7VSy+9pISEBMXHxys9PV0xMTHOJ7aSkpLUrVs3Pfnkk5o7d67Ky8uVlpam3r17V/pJLAAAYG1eDTsbN27UnXfe6fw8fPhwSVK/fv20YMECjRw5UsePH9eAAQNUXFysW2+9VcuWLVNgYKDzOwsXLlRaWpo6d+7sfKngzJkzq/1aAACAb/KZ9+x4E+/ZAQCg5qnx79kBAABwB8IOAACwNMIOAACwNMIOAACwNJ9dLgLWl5ubq6KiIo/1HxERodjYWI/1DwCoGQg78Irc3FwlJiaptPSEx84RFFRf27fnEHgAoJYj7MArioqKVFp6Qh0eH6uQ6Di39+/I26ust8erqKiIsAMAtRxhB14VEh2n8NgW3i4DAGBhTFAGAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACWRtgBAACW5uftAgBPysnJ8VjfERERio2N9Vj/AAD3IOzAkkpLDkmyqW/fvh47R1BQfW3fnkPgAQAfR9iBJZWfOCrJqM1Do3RVfKLb+3fk7VXW2+NVVFRE2AEAH0fYgaUFR8YqPLaFt8sAAHgRE5QBAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAIClEXYAAICl+XTYOXPmjNLT0xUfH6+goCA1b95cL774oowxzmOMMRozZoyio6MVFBSk1NRUff/9916sGgAA+BKfDjuTJ0/WnDlz9MYbbygnJ0eTJ0/WlClTNGvWLOcxU6ZM0cyZMzV37lxlZWWpQYMG6tq1q06ePOnFygEAgK/w83YBl/LVV1/p/vvvV/fu3SVJcXFxeu+997R+/XpJP93VmTFjhl544QXdf//9kqS///3vstvt+vjjj9W7d2+v1Q4AAHyDT9/ZueWWW5SRkaGdO3dKkr755hutXbtWd999tyRpz549ys/PV2pqqvM7oaGh6tChgzIzMy/ab1lZmRwOh8sGAACsyafv7Dz33HNyOBxKTExU3bp1debMGb388svq06ePJCk/P1+SZLfbXb5nt9ud+y5k4sSJGj9+vOcKBwAAPsOn7+y8//77WrhwoRYtWqRNmzbpnXfe0auvvqp33nnnivodPXq0SkpKnNv+/fvdVDEAAPA1Pn1nZ8SIEXruueecc29atWqlffv2aeLEierXr5+ioqIkSQUFBYqOjnZ+r6CgQG3atLlovwEBAQoICPBo7QAAwDf49J2dEydOqE4d1xLr1q2riooKSVJ8fLyioqKUkZHh3O9wOJSVlaWUlJRqrRUAAPgmn76zc++99+rll19WbGysWrZsqa+//lrTp0/X448/Lkmy2WwaOnSoXnrpJSUkJCg+Pl7p6emKiYlRjx49vFs8AADwCT4ddmbNmqX09HT98Y9/VGFhoWJiYvTUU09pzJgxzmNGjhyp48ePa8CAASouLtatt96qZcuWKTAw0IuVAwAAX+HTYadhw4aaMWOGZsyYcdFjbDabJkyYoAkTJlRfYQAAoMbw6Tk7AAAAV4qwAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALI2wAwAALM3P2wUANVlOTo7H+o6IiFBsbKzH+geA2oKwA1RBackhSTb17dvXY+cICqqv7dtzCDwAcIUIO0AVlJ84KsmozUOjdFV8otv7d+TtVdbb41VUVETYAYArRNgBrkBwZKzCY1t4uwwAwCUwQRkAAFgaYQcAAFgaYQcAAFgaYQcAAFgaYQcAAFgaYQcAAFgaYQcAAFgaYQcAAFgaYQcAAFgaYQcAAFgay0XgknJzc1VUVOT2fj25WjgAAD9H2MFF5ebmKjExSaWlJzx2jvKyUx7rGwAAibCDSygqKlJp6Ql1eHysQqLj3Np33reZ2vrJWzp9+rRb+wUA4FyEHfyikOg4t6/s7cjb69b+AAC4GCYoAwAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAASyPsAAAAS/P5sHPgwAH17dtXjRs3VlBQkFq1aqWNGzc69xtjNGbMGEVHRysoKEipqan6/vvvvVgxAADwJT4ddo4cOaJOnTrJ399fn332mbZt26Zp06apUaNGzmOmTJmimTNnau7cucrKylKDBg3UtWtXnTx50ouVAwAAX+Hn7QIuZfLkyWratKnmz5/vbIuPj3f+2RijGTNm6IUXXtD9998vSfr73/8uu92ujz/+WL179672mgEAgG/x6Ts7n3zyiZKTk/W73/1OkZGRatu2rebNm+fcv2fPHuXn5ys1NdXZFhoaqg4dOigzM9MbJQMAAB/j02Fn9+7dmjNnjhISEvT555/rmWee0eDBg/XOO+9IkvLz8yVJdrvd5Xt2u92570LKysrkcDhcNgAAYE0+/TNWRUWFkpOT9corr0iS2rZtq61bt2ru3Lnq169flfudOHGixo8f764yAQCAD/PpOzvR0dG6/vrrXdqSkpKUm5srSYqKipIkFRQUuBxTUFDg3Hcho0ePVklJiXPbv3+/mysHAAC+wqfDTqdOnbRjxw6Xtp07d6pZs2aSfpqsHBUVpYyMDOd+h8OhrKwspaSkXLTfgIAAhYSEuGwAAMCafPpnrGHDhumWW27RK6+8ot///vdav3693nrrLb311luSJJvNpqFDh+qll15SQkKC4uPjlZ6erpiYGPXo0cO7xQMAAJ/g02Gnffv2WrJkiUaPHq0JEyYoPj5eM2bMUJ8+fZzHjBw5UsePH9eAAQNUXFysW2+9VcuWLVNgYKAXKwcAAL7Cp8OOJN1zzz265557LrrfZrNpwoQJmjBhQjVWBQAAagqfnrMDAABwpQg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0qoUdq655hodOnTovPbi4mJdc801V1wUAACAu1Qp7Ozdu1dnzpw5r72srEwHDhy44qIAAADc5bKWi/jkk0+cf/78888VGhrq/HzmzBllZGQoLi7ObcUBAABcqcsKO2dXErfZbOrXr5/LPn9/f8XFxWnatGluKw4AAOBKXVbYqaiokCTFx8drw4YNioiI8EhRAAAA7lKlVc/37Nnj7joAAAA8okphR5IyMjKUkZGhwsJC5x2fs95+++0rLgwAAMAdqhR2xo8frwkTJig5OVnR0dGy2WzurgsAAMAtqhR25s6dqwULFujhhx92dz0AAABuVaX37Jw6dUq33HKLu2sBAABwuyqFnSeeeEKLFi1ydy0AAABuV6WfsU6ePKm33npLK1asUOvWreXv7++yf/r06W4pDgAA4EpVKexs2bJFbdq0kSRt3brVZR+TlQEAgC+pUtj54osv3F0HAACAR1Rpzg4AAEBNUaU7O3feeeclf65auXJllQsCAABwpyqFnbPzdc4qLy/X5s2btXXr1vMWCAUAAPCmKoWd11577YLt48aN07Fjx66oIAAAAHeq8tpYF9K3b1/dfPPNevXVV93ZLVBr5eTkeKTfiIgIxcbGeqRvAPA1bg07mZmZCgwMdGeXQK1UWnJIkk19+/b1SP9BQfW1fXsOgQdArVClsNOzZ0+Xz8YY5eXlaePGjUpPT3dLYUBtVn7iqCSjNg+N0lXxiW7t25G3V1lvj1dRURFhB0CtUKWwExoa6vK5Tp06atGihSZMmKAuXbq4pTAAUnBkrMJjW3i7DACo0aoUdubPn+/uOgAAADziiubsZGdnOydQtmzZUm3btnVLUQAAAO5SpbBTWFio3r17a9WqVQoLC5MkFRcX684779TixYt11VVXubNGAACAKqvSchGDBg3S0aNH9d133+nw4cM6fPiwtm7dKofDocGDB7u7RgAAgCqr0p2dZcuWacWKFUpKSnK2XX/99Zo9ezYTlAEAgE+p0p2diooK+fv7n9fu7++vioqKKy4KAADAXaoUdu666y4NGTJEBw8edLYdOHBAw4YNU+fOnd1WHAAAwJWqUth544035HA4FBcXp+bNm6t58+aKj4+Xw+HQrFmz3F0jAABAlVVpzk7Tpk21adMmrVixQtu3b5ckJSUlKTU11a3FAQAAXKnLurOzcuVKXX/99XI4HLLZbPr1r3+tQYMGadCgQWrfvr1atmypL7/80lO1AgAAXLbLCjszZszQk08+qZCQkPP2hYaG6qmnntL06dPdVhwAAMCVuqyw880336hbt24X3d+lSxdlZ2dfcVEAAADucllhp6Cg4IKPnJ/l5+enH3/88YqLAgAAcJfLCjtXX321tm7detH9W7ZsUXR09BUXBQAA4C6XFXZ+85vfKD09XSdPnjxvX2lpqcaOHat77rnHbcUBAABcqct69PyFF17QRx99pOuuu05paWlq0aKFJGn79u2aPXu2zpw5o+eff94jhQIAAFTFZYUdu92ur776Ss8884xGjx4tY4wkyWazqWvXrpo9e7bsdrtHCgUAAKiKy36pYLNmzfTvf/9bR44c0a5du2SMUUJCgho1auSJ+gAAAK5Ild6gLEmNGjVS+/bt3VkLAACA21VpbSwAAICagrADAAAsrco/YwGo2XJycjzWd0REhGJjYz3WPwBcDsIOUMuUlhySZFPfvn09do6goPravj2HwAPAJxB2gFqm/MRRSUZtHhqlq+IT3d6/I2+vst4er6KiIsIOAJ9A2AFqqeDIWIXHtvB2GQDgcUxQBgAAlkbYAQAAlkbYAQAAlkbYAQAAlkbYAQAAlkbYAQAAlkbYAQAAllajws6kSZNks9k0dOhQZ9vJkyc1cOBANW7cWMHBwerVq5cKCgq8VyQAAPApNSbsbNiwQX/5y1/UunVrl/Zhw4bp008/1QcffKDVq1fr4MGD6tmzp5eqBAAAvqZGhJ1jx46pT58+mjdvnho1auRsLykp0d/+9jdNnz5dd911l9q1a6f58+frq6++0rp167xYMQAA8BU1IuwMHDhQ3bt3V2pqqkt7dna2ysvLXdoTExMVGxurzMzMi/ZXVlYmh8PhsgEAAGvy+bWxFi9erE2bNmnDhg3n7cvPz1e9evUUFhbm0m6325Wfn3/RPidOnKjx48e7u1QAAOCDfPrOzv79+zVkyBAtXLhQgYGBbut39OjRKikpcW779+93W98AAMC3+HTYyc7OVmFhoW666Sb5+fnJz89Pq1ev1syZM+Xn5ye73a5Tp06puLjY5XsFBQWKioq6aL8BAQEKCQlx2QAAgDX59M9YnTt31rfffuvS9thjjykxMVGjRo1S06ZN5e/vr4yMDPXq1UuStGPHDuXm5iolJcUbJQMAAB/j02GnYcOGuuGGG1zaGjRooMaNGzvb+/fvr+HDhys8PFwhISEaNGiQUlJS1LFjR2+UDAAAfIxPh53KeO2111SnTh316tVLZWVl6tq1q958801vlwUAAHxEjQs7q1atcvkcGBio2bNna/bs2d4pCAAA+DSfnqAMAABwpQg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0gg7AADA0vy8XQCuTG5uroqKijzSd05Ojkf6BQCgOhF2arDc3FwlJiaptPSER89TXnbKo/0DAOBJhJ0arKioSKWlJ9Th8bEKiY5ze/9532Zq6ydv6fTp027vGwCA6kLYsYCQ6DiFx7Zwe7+OvL1u7xMAgOrGBGUAAGBphB0AAGBphB0AAGBpzNkB4BGefHVBRESEYmNjPdY/AGsh7ABwq9KSQ5Js6tu3r8fOERRUX9u35xB4AFQKYQeAW5WfOCrJqM1Do3RVfKLb+3fk7VXW2+NVVFRE2AFQKYQdAB4RHBnrkVciAMDlYoIyAACwNMIOAACwNMIOAACwNMIOAACwNMIOAACwNMIOAACwNMIOAACwNMIOAACwNMIOAACwNMIOAACwNMIOAACwNNbGAoBz5ObmqqioyCN9R0REsIApUM0IOwDwM7m5uUpMTFJp6QmP9B8UVF/bt+cQeIBqRNgBgJ8pKipSaekJdXh8rEKi49zatyNvr7LeHq+ioiLCDlCNCDsAcAEh0XEKj23h7TIAuIFPT1CeOHGi2rdvr4YNGyoyMlI9evTQjh07XI45efKkBg4cqMaNGys4OFi9evVSQUGBlyoGAAC+xqfDzurVqzVw4ECtW7dOy5cvV3l5ubp06aLjx487jxk2bJg+/fRTffDBB1q9erUOHjyonj17erFqAADgS3z6Z6xly5a5fF6wYIEiIyOVnZ2tX/3qVyopKdHf/vY3LVq0SHfddZckaf78+UpKStK6devUsWNHb5QNAAB8iE/f2TlXSUmJJCk8PFySlJ2drfLycqWmpjqPSUxMVGxsrDIzMy/aT1lZmRwOh8sGAACsqcaEnYqKCg0dOlSdOnXSDTfcIEnKz89XvXr1FBYW5nKs3W5Xfn7+RfuaOHGiQkNDnVvTpk09WToAAPCiGhN2Bg4cqK1bt2rx4sVX3Nfo0aNVUlLi3Pbv3++GCgEAgC/y6Tk7Z6WlpWnp0qVas2aNmjRp4myPiorSqVOnVFxc7HJ3p6CgQFFRURftLyAgQAEBAZ4sGQAA+AifvrNjjFFaWpqWLFmilStXKj4+3mV/u3bt5O/vr4yMDGfbjh07lJubq5SUlOouFwAA+CCfvrMzcOBALVq0SP/85z/VsGFD5zyc0NBQBQUFKTQ0VP3799fw4cMVHh6ukJAQDRo0SCkpKTyJBQAAJPl42JkzZ44k6Y477nBpnz9/vh599FFJ0muvvaY6deqoV69eKisrU9euXfXmm29Wc6UAAMBX+XTYMcb84jGBgYGaPXu2Zs+eXQ0VAcCVy8nJ8VjfrKoOnM+nww4AWElpySFJNvXt29dj52BVdeB8hB0AqCblJ45KMmrz0ChdFZ/o9v5ZVR24MMIOAFSz4MhYVlQHqpFPP3oOAABwpQg7AADA0gg7AADA0gg7AADA0pigDKBG8tS7ajz5DhwA3kHYAVCjVMe7aiSpvOyUR/sHUH0IOwBqFE+/qybv20xt/eQtnT592u19A/AOwg6AGslT76px5O11e5/VjeUoAFeEHQCwCJajAC6MsAMAFsFyFMCFEXYAwGJYjgJwRdjxsNzcXBUVFXmkbx6RBQDglxF2PCg3N1eJiUkqLT3h0fPwiCwAABdH2PGgoqIilZaeUIfHxyokOs7t/fOILAAAv4ywUw1CouN4RBYAAC9hbSwAAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBphB0AAGBpft4uAABQs+Tk5Hik34iICMXGxnqkb9RuhB0AQKWUlhySZFPfvn090n9QUH1t355D4IHbEXYAAJVSfuKoJKM2D43SVfGJbu3bkbdXWW+PV1FREWEHbkfYAQBcluDIWIXHtvB2GUClMUEZAABYGmEHAABYGmEHAABYGmEHAABYGmEHAABYGmEHAABYGmEHAABYGu/ZAQD4DE8tRSGxHEVtRtgBAHidp5eikFiOojYj7AAAvM6TS1FILEdR2xF2AAA+g6Uo4AmEHQBArcGcoNqJsAMAsDzmBNVuhB0AgOUxJ6h2I+wAAGoN5gTVTrxUEAAAWBphBwAAWBphBwAAWJpl5uzMnj1bU6dOVX5+vm688UbNmjVLN998s7fLAgDUIp58tL2srEwBAQE1rm/J+4/lWyLs/OMf/9Dw4cM1d+5cdejQQTNmzFDXrl21Y8cORUZGers8AIDFVcej7bLZJGNqXt/y/mP5lgg706dP15NPPqnHHntMkjR37lz961//0ttvv63nnnvOy9UBAKzO04+2532bqa2fvOWR/j3Zt+Qbj+XX+LBz6tQpZWdna/To0c62OnXqKDU1VZmZmV6sDABQ23jq0XZH3l6P9e/Jvn1FjQ87RUVFOnPmjOx2u0u73W7X9u3bL/idsrIylZWVOT+XlJRIkhwOh1trO3bsmCTp8L4dOl1W6ta+JcmRt0+SVHLge/n72WpU/zW59pref02uvab3X5Nr93T/Nbn2mt6/x2vPz5X007+J7v539mx/5pd+gjM13IEDB4wk89VXX7m0jxgxwtx8880X/M7YsWONJDY2NjY2NjYLbPv3779kVqjxd3YiIiJUt25dFRQUuLQXFBQoKirqgt8ZPXq0hg8f7vxcUVGhw4cPq3HjxrLZ3J9qL5fD4VDTpk21f/9+hYSEeLscr2M8zseYuGI8XDEerhiP81llTIwxOnr0qGJiYi55XI0PO/Xq1VO7du2UkZGhHj16SPopvGRkZCgtLe2C3wkICDjvEbuwsDAPV3r5QkJCavT/hO7GeJyPMXHFeLhiPFwxHuezwpiEhob+4jE1PuxI0vDhw9WvXz8lJyfr5ptv1owZM3T8+HHn01kAAKD2skTY+cMf/qAff/xRY8aMUX5+vtq0aaNly5adN2kZAADUPpYIO5KUlpZ20Z+tapqAgACNHTvWo2+zrEkYj/MxJq4YD1eMhyvG43y1bUxsxnjwlYkAAABexkKgAADA0gg7AADA0gg7AADA0gg7XjJx4kS1b99eDRs2VGRkpHr06KEdO3a4HHPy5EkNHDhQjRs3VnBwsHr16nXeyxOtatKkSbLZbBo6dKizrTaOx4EDB9S3b181btxYQUFBatWqlTZu3Ojcb4zRmDFjFB0draCgIKWmpur777/3YsWec+bMGaWnpys+Pl5BQUFq3ry5XnzxRZfXxFt9PNasWaN7771XMTExstls+vjjj132V+b6Dx8+rD59+igkJERhYWHq37+/c2mbmuZS41FeXq5Ro0apVatWatCggWJiYvTII4/o4MGDLn3UlvE419NPPy2bzaYZM2a4tFtpPH6OsOMlq1ev1sCBA7Vu3TotX75c5eXl6tKli44fP+48ZtiwYfr000/1wQcfaPXq1Tp48KB69uzpxaqrx4YNG/SXv/xFrVu3dmmvbeNx5MgRderUSf7+/vrss8+0bds2TZs2TY0aNXIeM2XKFM2cOVNz585VVlaWGjRooK5du+rkyZNerNwzJk+erDlz5uiNN95QTk6OJk+erClTpmjWrFnOY6w+HsePH9eNN96o2bNnX3B/Za6/T58++u6777R8+XItXbpUa9as0YABA6rrEtzqUuNx4sQJbdq0Senp6dq0aZM++ugj7dixQ/fdd5/LcbVlPH5uyZIlWrdu3QXfOmyl8XBx5atTwR0KCwuNJLN69WpjjDHFxcXG39/ffPDBB85jcnJyjCSTmZnprTI97ujRoyYhIcEsX77c3H777WbIkCHGmNo5HqNGjTK33nrrRfdXVFSYqKgoM3XqVGdbcXGxCQgIMO+99151lFitunfvbh5//HGXtp49e5o+ffoYY2rfeEgyS5YscX6uzPVv27bNSDIbNmxwHvPZZ58Zm81mDhw4UG21e8K543Eh69evN5LMvn37jDG1czz+97//mauvvtps3brVNGvWzLz22mvOfVYeD+7s+IizK6+Hh4dLkrKzs1VeXq7U1FTnMYmJiYqNjVVmZqZXaqwOAwcOVPfu3V2uW6qd4/HJJ58oOTlZv/vd7xQZGam2bdtq3rx5zv179uxRfn6+y5iEhoaqQ4cOlhyTW265RRkZGdq5c6ck6ZtvvtHatWt19913S6p943Guylx/ZmamwsLClJyc7DwmNTVVderUUVZWVrXXXN1KSkpks9mcywPVtvGoqKjQww8/rBEjRqhly5bn7bfyeFjmpYI1WUVFhYYOHapOnTrphhtukCTl5+erXr16563ZZbfblZ+f74UqPW/x4sXatGmTNmzYcN6+2jgeu3fv1pw5czR8+HD9+c9/1oYNGzR48GDVq1dP/fr1c173uW8Kt+qYPPfcc3I4HEpMTFTdunV15swZvfzyy+rTp48k1brxOFdlrj8/P1+RkZEu+/38/BQeHm75MTp58qRGjRqlBx980LkWVG0bj8mTJ8vPz0+DBw++4H4rjwdhxwcMHDhQW7du1dq1a71ditfs379fQ4YM0fLlyxUYGOjtcnxCRUWFkpOT9corr0iS2rZtq61bt2ru3Lnq16+fl6urfu+//74WLlyoRYsWqWXLltq8ebOGDh2qmJiYWjkeqLzy8nL9/ve/lzFGc+bM8XY5XpGdna3XX39dmzZtks1m83Y51Y6fsbwsLS1NS5cu1RdffKEmTZo426OionTq1CkVFxe7HF9QUKCoqKhqrtLzsrOzVVhYqJtuukl+fn7y8/PT6tWrNXPmTPn5+clut9eq8ZCk6OhoXX/99S5tSUlJys3NlSTndZ/7RJpVx2TEiBF67rnn1Lt3b7Vq1UoPP/ywhg0bpokTJ0qqfeNxrspcf1RUlAoLC132nz59WocPH7bsGJ0NOvv27dPy5ctdVviuTePx5ZdfqrCwULGxsc6/Y/ft26dnn31WcXFxkqw9HoQdLzHGKC0tTUuWLNHKlSsVHx/vsr9du3by9/dXRkaGs23Hjh3Kzc1VSkpKdZfrcZ07d9a3336rzZs3O7fk5GT16dPH+efaNB6S1KlTp/NeR7Bz5041a9ZMkhQfH6+oqCiXMXE4HMrKyrLkmJw4cUJ16rj+lVW3bl1VVFRIqn3jca7KXH9KSoqKi4uVnZ3tPGblypWqqKhQhw4dqr1mTzsbdL7//nutWLFCjRs3dtlfm8bj4Ycf1pYtW1z+jo2JidGIESP0+eefS7L4eHh7hnRt9cwzz5jQ0FCzatUqk5eX59xOnDjhPObpp582sbGxZuXKlWbjxo0mJSXFpKSkeLHq6vXzp7GMqX3jsX79euPn52defvll8/3335uFCxea+vXrm3fffdd5zKRJk0xYWJj55z//abZs2WLuv/9+Ex8fb0pLS71YuWf069fPXH311Wbp0qVmz5495qOPPjIRERFm5MiRzmOsPh5Hjx41X3/9tfn666+NJDN9+nTz9ddfO58uqsz1d+vWzbRt29ZkZWWZtWvXmoSEBPPggw9665KuyKXG49SpU+a+++4zTZo0MZs3b3b5e7asrMzZR20Zjws592ksY6w1Hj9H2PESSRfc5s+f7zymtLTU/PGPfzSNGjUy9evXNw888IDJy8vzXtHV7NywUxvH49NPPzU33HCDCQgIMImJieatt95y2V9RUWHS09ON3W43AQEBpnPnzmbHjh1eqtazHA6HGTJkiImNjTWBgYHmmmuuMc8//7zLP1xWH48vvvjign9v9OvXzxhTues/dOiQefDBB01wcLAJCQkxjz32mDl69KgXrubKXWo89uzZc9G/Z7/44gtnH7VlPC7kQmHHSuPxc6x6DgAALI05OwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIOwAAwNIIO0ANdMcdd2jo0KEePcejjz6qHj16/OJxNptNH3/8sUdr8SXVMfZVtXfvXtlsNm3evNnbpQA+hbAD4Irk5eXp7rvvlsQ/tgB8E2EHQJWcOnVKkhQVFaWAgAAvV+Pbzo4VAO8g7AA+7vjx43rkkUcUHBys6OhoTZs2zWV/WVmZ/vSnP+nqq69WgwYN1KFDB61atcq5f8GCBQoLC9Pnn3+upKQkBQcHq1u3bsrLy3Mec+bMGQ0fPlxhYWFq3LixRo4cqXOXzbvjjjuUlpamoUOHKiIiQl27dpXk+jNWfHy8JKlt27ay2Wy64447nN//61//qqSkJAUGBioxMVFvvvnmJa/7jjvu0ODBgzVy5EiFh4crKipK48aNc+6/0F2k4uJi2Ww25/WvWrVKNptNn3/+udq2baugoCDdddddKiws1GeffaakpCSFhITooYce0okTJy5Zz8+dPn1aaWlpCg0NVUREhNLT013GKy4uTi+++KIeeeQRhYSEaMCAAZKkUaNG6brrrlP9+vV1zTXXKD09XeXl5c7vjRs3Tm3atNH//d//KS4uTqGhoerdu7eOHj3qPKaiokJTpkzRtddeq4CAAMXGxurll192qW/37t268847Vb9+fd14443KzMx02b927VrddtttCgoKUtOmTTV48GAdP37cuf/NN99UQkKCAgMDZbfb9dvf/rbSYwP4JO+uQwrglzzzzDMmNjbWrFixwmzZssXcc889pmHDhs4V4Z944glzyy23mDVr1phdu3aZqVOnmoCAALNz505jjDHz5883/v7+JjU11WzYsMFkZ2ebpKQk89BDDznPMXnyZNOoUSPz4Ycfmm3btpn+/fubhg0bmvvvv995zO23326Cg4PNiBEjzPbt28327duNMcZIMkuWLDHGGLN+/XojyaxYscLk5eWZQ4cOGWOMeffdd010dLT58MMPze7du82HH35owsPDzYIFCy563bfffrsJCQkx48aNMzt37jTvvPOOsdls5j//+Y8xxjhXtf7666+d3zly5IjLqtZnV4Hu2LGjWbt2rdm0aZO59tprze233266dOliNm3aZNasWWMaN25sJk2aVKn/HmfHYciQIWb79u3m3XffNfXr13dZkb5Zs2YmJCTEvPrqq2bXrl1m165dxhhjXnzxRfPf//7X7Nmzx3zyySfGbrebyZMnO783duxYExwcbHr27Gm+/fZbs2bNGhMVFWX+/Oc/O48ZOXKkadSokVmwYIHZtWuX+fLLL828efNcxiQxMdEsXbrU7Nixw/z2t781zZo1M+Xl5cYYY3bt2mUaNGhgXnvtNbNz507z3//+17Rt29Y8+uijxhhjNmzYYOrWrWsWLVpk9u7dazZt2mRef/31So0N4KsIO4APO3r0qKlXr555//33nW2HDh0yQUFBZsiQIWbfvn2mbt265sCBAy7f69y5sxk9erQx5qewI8n5D64xxsyePdvY7Xbn5+joaDNlyhTn5/LyctOkSZPzwk7btm3Pq/HnYedCAcQYY5o3b24WLVrk0vbiiy+alJSUi1777bffbm699VaXtvbt25tRo0Zd9FwXCzsrVqxwHjNx4kQjyfzwww/Otqeeesp07dr1orWcW1dSUpKpqKhwto0aNcokJSU5Pzdr1sz06NHjF/uaOnWqadeunfPz2LFjTf369Y3D4XC2jRgxwnTo0MEYY4zD4TABAQHOcHOus2Py17/+1dn23XffGUkmJyfHGGNM//79zYABA1y+9+WXX5o6deqY0tJS8+GHH5qQkBCXGoCazs8LN5MAVNIPP/ygU6dOqUOHDs628PBwtWjRQpL07bff6syZM7ruuutcvldWVqbGjRs7P9evX1/Nmzd3fo6OjlZhYaEkqaSkRHl5eS7n8PPzU3Jy8nk/ZbVr1+6yr+H48eP64Ycf1L9/fz355JPO9tOnTys0NPSS323durXL55/XfTl+3o/dbnf+jPTztvXr11e6v44dO8pmszk/p6SkaNq0aTpz5ozq1q0rSUpOTj7ve//4xz80c+ZM/fDDDzp27JhOnz6tkJAQl2Pi4uLUsGFD5+efX3NOTo7KysrUuXPnSl9vdHS0JKmwsFCJiYn65ptvtGXLFi1cuNB5jDFGFRUV2rNnj37961+rWbNmuuaaa9StWzd169ZNDzzwgOrXr1/Z4QF8DmEHqMGOHTumunXrKjs72/mP7FnBwcHOP/v7+7vss9ls5wWZymjQoEGVapSkefPmuQQqSefVfK4L1V1RUSFJqlPnpymHP7+On89/uVg/Npvtkv26y7ljlZmZqT59+mj8+PHq2rWrQkNDtXjx4vPmYF2qtqCgoEqd+9zrleTs49ixY3rqqac0ePDg874XGxurevXqadOmTVq1apX+85//aMyYMRo3bpw2bNigsLCwSp0f8DVMUAZ8WPPmzeXv76+srCxn25EjR7Rz505JP00EPnPmjAoLC3Xttde6bFFRUZU6R2hoqKKjo13Ocfr0aWVnZ192vfXq1ZP004Tns+x2u2JiYrR79+7zajw7obkqrrrqKklymWhdXY+8/3ysJGndunVKSEi4ZHj76quv1KxZMz3//PNKTk5WQkKC9u3bd1nnTUhIUFBQkDIyMqpUtyTddNNN2rZt23n/La699lrnfz8/Pz+lpqZqypQp2rJli/bu3auVK1dW+ZyAt3FnB/BhwcHB6t+/v0aMGKHGjRsrMjJSzz//vPOuxnXXXac+ffrokUce0bRp09S2bVv9+OOPysjIUOvWrdW9e/dKnWfIkCGaNGmSEhISlJiYqOnTp6u4uPiy642MjFRQUJCWLVumJk2aKDAwUKGhoRo/frwGDx6s0NBQdevWTWVlZdq4caOOHDmi4cOHS5I6d+6sBx54QGlpaZU6V1BQkDp27KhJkyYpPj5ehYWFeuGFFy675qrIzc3V8OHD9dRTT2nTpk2aNWvWeXdozpWQkKDc3FwtXrxY7du317/+9S8tWbLkss4bGBioUaNGaeTIkapXr546deqkH3/8Ud9995369+9fqT5GjRqljh07Ki0tTU888YQaNGigbdu2afny5XrjjTe0dOlS7d69W7/61a/UqFEj/fvf/1ZFRYXzp1OgJiLsAD5u6tSpOnbsmO699141bNhQzz77rEpKSpz758+fr5deeknPPvusDhw4oIiICHXs2FH33HNPpc/x7LPPKi8vT/369VOdOnX0+OOP64EHHnA5T2X4+flp5syZmjBhgsaMGaPbbrtNq1at0hNPPKH69etr6tSpGjFihBo0aKBWrVq5vIn4hx9+UFFR0WWd7+2331b//v3Vrl07tWjRQlOmTFGXLl0uq4+qeOSRR1RaWqqbb75ZdevW1ZAhQ5yPl1/Mfffdp2HDhiktLU1lZWXq3r270tPTXR6nr4z09HT5+flpzJgxOnjwoKKjo/X0009X+vutW7fW6tWr9fzzz+u2226TMUbNmzfXH/7wB0lSWFiYPvroI40bN04nT55UQkKC3nvvPbVs2fKy6gR8ic1U5Yd7AACAGoI5OwAAwNIIOwDw/+Xm5io4OPiiW25urrdLBFAF/IwFAP/f6dOntXfv3ovuj4uLk58fUx2BmoawAwAALI2fsQAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKURdgAAgKX9P5mvJI4Kh8EHAAAAAElFTkSuQmCC\n"
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, we should determine if there is NA or missing data using `.isna()` and `.any()`."
      ],
      "metadata": {
        "id": "7fRsjX20pUKS"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cell_data[\"dendrite.num_branches\"].isna().any()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "b1xdiEFxgyH9",
        "outputId": "fdf2b0fe-76f7-48ee-b583-6586dfa177bd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "False"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We can see there is no missing data for dendrite.num_branches; however, we should also verify this for the gene expression data, which can be accessed using `cell_data.iloc[:, :-24]`."
      ],
      "metadata": {
        "id": "QLF8PvL7iqRG"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cell_data.iloc[:, :-24].isna().any().any()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ZS2KvhvtiklA",
        "outputId": "4e578861-d76b-49b7-fb4e-6ad8f6a91abd"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "False"
            ]
          },
          "metadata": {},
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "We should also examine the metadata, specifically neuron_reconstruction_type, to check the morphological data for any anomalies. First we combine `cell_data` and `cell_metadata` to get the metadata for the specific neurons in `cell_data`, as `cell_metadata` includes data for other neurons without morphological reconstructions. Printing the columns shows that the metadata is now included."
      ],
      "metadata": {
        "id": "UctQybaak9ft"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "cell_data = cell_data.join(cell_metadata, how=\"inner\")\n",
        "cell_data.columns"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "il-3N62kmSZe",
        "outputId": "d3ec87ee-f672-49a6-cc3c-214e24ed7fec"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['0610005C13Rik', '0610006L08Rik', '0610007P14Rik', '0610009B22Rik',\n",
              "       '0610009E02Rik', '0610009L18Rik', '0610009O20Rik', '0610010B08Rik',\n",
              "       '0610010F05Rik', '0610010K14Rik',\n",
              "       ...\n",
              "       'apical_dendrite_status', 'neuron_reconstruction_type',\n",
              "       'cell_soma_normalized_depth', 'depth_from_pia_um', 'ephys_session_id',\n",
              "       'transcriptomics_sample_id', 'transcriptomics_batch',\n",
              "       'T-type Accession ID', 'T-type Label', 'MET-type Label'],\n",
              "      dtype='object', length=45814)"
            ]
          },
          "metadata": {},
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "cell_data[\"neuron_reconstruction_type\"].unique()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "z0bUBHAYq3hr",
        "outputId": "91343dc3-dd43-46ea-ef28-5d0089a3d974"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['full', 'dendrite-only'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "By looking at the different reconstruction types, we can see that some neurons have reconstructions of the dendrites only. Because we are focusing on dendritic morphology, this will not affect the analysis. However, if you intend to analyze an axon feature, you should remove the cells with incomplete reconstructions."
      ],
      "metadata": {
        "id": "nWutS2Mbrcde"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "##GSEA Setup: The Ranking Metric and Ranked List of Genes"
      ],
      "metadata": {
        "id": "yb9hamcQsnbw"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "The first step of GSEA requires a list of all the genes in the dataset ranked by a suitable metric. There are multiple different metrics that can be used depending on the type of data, but the metric should ultimately reflect the correlation between a gene's expression level and the phenotype. GSEA is often applied to datasets with two distinct classes or experimental conditions (e.g. normal cells vs. disease state cells) to identify pathways that are upregulated in one class/condition compared to the other. In that case, a metric incorporating the differential expression of each gene (fold change) between the two classes is used. For this analysis, we are analyzing a continuous phenotype where there are not two distinct classes. Therefore, we will use the Pearson correlation coefficent as our ranking metric instead."
      ],
      "metadata": {
        "id": "dHfw8_jgs1lP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "corrs = cell_data.iloc[:, :-46].corrwith(cell_data[\"dendrite.num_branches\"]) #Computing correlations between each gene with dendritic branching.\n",
        "corrs = corrs.sort_values(ascending=False) #Ranking the list with the most correlated genes at the top.\n",
        "corrs.head(7)"
      ],
      "metadata": {
        "id": "DTaPGn7ipeBm",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 373
        },
        "outputId": "295e05a3-8e5e-48b2-a254-46cc58702c83"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/numpy/lib/function_base.py:2897: RuntimeWarning: invalid value encountered in divide\n",
            "  c /= stddev[:, None]\n",
            "/usr/local/lib/python3.11/dist-packages/numpy/lib/function_base.py:2898: RuntimeWarning: invalid value encountered in divide\n",
            "  c /= stddev[None, :]\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Arpp21      0.512190\n",
              "Slc17a7     0.508691\n",
              "Cacna2d1    0.499428\n",
              "Ptprd       0.491913\n",
              "R3hdm1      0.487666\n",
              "Phactr1     0.482130\n",
              "Nell2       0.480631\n",
              "dtype: float64"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>Arpp21</th>\n",
              "      <td>0.512190</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Slc17a7</th>\n",
              "      <td>0.508691</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Cacna2d1</th>\n",
              "      <td>0.499428</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Ptprd</th>\n",
              "      <td>0.491913</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>R3hdm1</th>\n",
              "      <td>0.487666</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Phactr1</th>\n",
              "      <td>0.482130</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Nell2</th>\n",
              "      <td>0.480631</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div><br><label><b>dtype:</b> float64</label>"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "corrs.tail(7)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 303
        },
        "id": "IsiN9zvId3cF",
        "outputId": "8a1f4592-08d7-42f3-8e5c-cfaf94e3c7f8"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "n-R5s139   NaN\n",
              "n-R5s141   NaN\n",
              "n-R5s142   NaN\n",
              "n-R5s143   NaN\n",
              "n-R5s144   NaN\n",
              "n-R5s146   NaN\n",
              "n-R5s149   NaN\n",
              "dtype: float64"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>n-R5s139</th>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>n-R5s141</th>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>n-R5s142</th>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>n-R5s143</th>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>n-R5s144</th>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>n-R5s146</th>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>n-R5s149</th>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div><br><label><b>dtype:</b> float64</label>"
            ]
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "As we can see by printing the tail of `corrs`, some genes have NaN correlation values. This is because these genes are not expressed in any sample in the dataset, and thus there is no spread in the expression data (standard deviation = 0), resulting in an undefined value. These genes will not be relevant for the analysis and can be removed from the ranked list `corrs`."
      ],
      "metadata": {
        "id": "T0ugizyHWGGc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "corrs = corrs.dropna()\n",
        "corrs.tail(7)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 303
        },
        "id": "kh4de8nwXvCX",
        "outputId": "cdbd9d6a-7053-448b-ecc1-b6639758da3b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Ube3a    -0.296282\n",
              "Dpp6     -0.299590\n",
              "Rab3b    -0.300480\n",
              "Elavl2   -0.300543\n",
              "Grip1    -0.303891\n",
              "Kcnc2    -0.304194\n",
              "Gria1    -0.315897\n",
              "dtype: float64"
            ],
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>Ube3a</th>\n",
              "      <td>-0.296282</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Dpp6</th>\n",
              "      <td>-0.299590</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Rab3b</th>\n",
              "      <td>-0.300480</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Elavl2</th>\n",
              "      <td>-0.300543</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Grip1</th>\n",
              "      <td>-0.303891</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Kcnc2</th>\n",
              "      <td>-0.304194</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Gria1</th>\n",
              "      <td>-0.315897</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div><br><label><b>dtype:</b> float64</label>"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Gene Set Enrichment Analysis (GSEA)"
      ],
      "metadata": {
        "id": "U3aG1MwnuXaW"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Selecting a Gene Set Database and Converting Gene Symbols"
      ],
      "metadata": {
        "id": "NAq4xY-kcT89"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Next, we need to select gene set databases to analyze. GSEApy\n"
      ],
      "metadata": {
        "id": "GDyF1dzqvFyw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "gp.get_library_name(organism=\"Mouse\") #Shows all gene set databases in GSEApy package\n"
      ],
      "metadata": {
        "id": "yeBbpTAouduJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from gseapy import Msigdb\n",
        "msig = Msigdb()\n",
        "# mouse hallmark gene sets\n",
        "gmt = msig.get_gmt(category='mh.all', dbver=\"2024.1.Mm\")"
      ],
      "metadata": {
        "id": "Ruoe5aK-dTzW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "\n",
        "# list categories given dbver.\n",
        "msig.list_category(dbver=\"2024.1.Mm\") # mouse\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rEcMm5K2doiO",
        "outputId": "cb347d99-f0ed-4196-9528-f2ba5c3c8009"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['m1.all',\n",
              " 'm2.all',\n",
              " 'm2.cgp',\n",
              " 'm2.cp.biocarta',\n",
              " 'm2.cp.reactome',\n",
              " 'm2.cp',\n",
              " 'm2.cp.wikipathways',\n",
              " 'm3.all',\n",
              " 'm3.gtrd',\n",
              " 'm3.mirdb',\n",
              " 'm5.all',\n",
              " 'm5.go.bp',\n",
              " 'm5.go.cc',\n",
              " 'm5.go.mf',\n",
              " 'm5.go',\n",
              " 'm5.mpt',\n",
              " 'm8.all',\n",
              " 'mh.all',\n",
              " 'msigdb']"
            ]
          },
          "metadata": {},
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# list msigdb version you wanna query\n",
        "msig.list_dbver()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 613
        },
        "id": "8ZrZOI6AdoF_",
        "outputId": "56695a1a-1fa2-4e8a-cd2a-e199c8b978ce"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "         Name     Last modified\n",
              "0         6.0  2018-12-17 16:29\n",
              "1         6.1  2018-12-17 16:29\n",
              "2         6.2  2018-12-17 16:29\n",
              "3         7.0  2020-01-27 04:43\n",
              "4         7.1  2020-03-30 16:55\n",
              "5         7.2  2020-09-18 23:33\n",
              "6         7.3  2021-03-22 15:09\n",
              "7         7.4  2021-04-02 14:38\n",
              "8       7.5.1  2023-03-09 19:38\n",
              "9         7.5  2022-01-13 01:00\n",
              "10  2022.1.Hs  2023-03-09 19:37\n",
              "11  2022.1.Mm  2023-03-09 19:37\n",
              "12  2023.1.Hs  2023-08-19 15:08\n",
              "13  2023.1.Mm  2023-08-19 15:09\n",
              "14  2023.2.Hs  2023-10-21 01:29\n",
              "15  2023.2.Mm  2023-10-20 08:32\n",
              "16  2024.1.Hs  2024-09-03 14:38\n",
              "17  2024.1.Mm  2024-09-03 14:38"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-f9590eb4-4b73-4eec-b132-b7419511e217\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Name</th>\n",
              "      <th>Last modified</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>6.0</td>\n",
              "      <td>2018-12-17 16:29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>6.1</td>\n",
              "      <td>2018-12-17 16:29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>6.2</td>\n",
              "      <td>2018-12-17 16:29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>7.0</td>\n",
              "      <td>2020-01-27 04:43</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>7.1</td>\n",
              "      <td>2020-03-30 16:55</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>7.2</td>\n",
              "      <td>2020-09-18 23:33</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>7.3</td>\n",
              "      <td>2021-03-22 15:09</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>7.4</td>\n",
              "      <td>2021-04-02 14:38</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>7.5.1</td>\n",
              "      <td>2023-03-09 19:38</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>7.5</td>\n",
              "      <td>2022-01-13 01:00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>2022.1.Hs</td>\n",
              "      <td>2023-03-09 19:37</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>2022.1.Mm</td>\n",
              "      <td>2023-03-09 19:37</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>2023.1.Hs</td>\n",
              "      <td>2023-08-19 15:08</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>2023.1.Mm</td>\n",
              "      <td>2023-08-19 15:09</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>2023.2.Hs</td>\n",
              "      <td>2023-10-21 01:29</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>2023.2.Mm</td>\n",
              "      <td>2023-10-20 08:32</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>2024.1.Hs</td>\n",
              "      <td>2024-09-03 14:38</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>2024.1.Mm</td>\n",
              "      <td>2024-09-03 14:38</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-f9590eb4-4b73-4eec-b132-b7419511e217')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-f9590eb4-4b73-4eec-b132-b7419511e217 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-f9590eb4-4b73-4eec-b132-b7419511e217');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-0ac376cd-8cc4-42b0-ac7c-d46628fccaa3\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-0ac376cd-8cc4-42b0-ac7c-d46628fccaa3')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-0ac376cd-8cc4-42b0-ac7c-d46628fccaa3 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"msig\",\n  \"rows\": 18,\n  \"fields\": [\n    {\n      \"column\": \"Name\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 18,\n        \"samples\": [\n          \"6.0\",\n          \"6.1\",\n          \"7.5.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Last modified\",\n      \"properties\": {\n        \"dtype\": \"object\",\n        \"num_unique_values\": 14,\n        \"samples\": [\n          \"2023-08-19 15:08\",\n          \"2023-10-21 01:29\",\n          \"2018-12-17 16:29\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "Mt_bOdjUdThX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#o make sure our gene symbols are correct for the database(s) we will later input into the GSEA algorithm."
      ],
      "metadata": {
        "id": "mVrzgkHKcGeQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#geneData.columns = geneData.columns.str.upper() #Making all gene identifiers uppercase for the correct gseapy/enrichr format (NOTE: Only if using the databases attached to gseapy/enrichr, which can be found with this gp.get_library_name(organism=\"Mouse\"))"
      ],
      "metadata": {
        "id": "DrFIO3uIcPkK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Reactome"
      ],
      "metadata": {
        "id": "LFxF7zaJro0J"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "LaW24D8Jrv1P"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##KEGG"
      ],
      "metadata": {
        "id": "lpuJMja4rqXe"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "tWZc7pR1rwcU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Gene Ontology (GO) Biological Process"
      ],
      "metadata": {
        "id": "DixdhVeXrykt"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "bDwzwN1Mr4tc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##GO Molecular Function"
      ],
      "metadata": {
        "id": "4lbcBhKKr5IB"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "LnAIqwgXsDSR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##GO Cellular Component"
      ],
      "metadata": {
        "id": "RW3hLwlosHDX"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "nA2jkn1rsKlx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Visualizations and Graphs"
      ],
      "metadata": {
        "id": "SKyJNQEXrgE7"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "##GSEA Plots"
      ],
      "metadata": {
        "id": "fWZudAthu2HX"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "cYH8M0QesV77"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Network Visualizations"
      ],
      "metadata": {
        "id": "v7Bob7jdu49r"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "8tRKRqO9u-7G"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Leading Edge Subset Analysis"
      ],
      "metadata": {
        "id": "DTvee-5DsWZZ"
      }
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "5FW-Kp9lsdlZ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Works Cited"
      ],
      "metadata": {
        "id": "jI-OpcJphvcc"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Aleksander, S., Balhoff, J. P., Carbon, S., J. Michael Cherry, Drabkin, H. J., Ebert, D., Feuermann, M., Gaudet, P., Harris, N. L., Hill, D. P., Lee, R., Mi, H., Sierra, Mungall, C. J., Anushya Muruganujan, Tremayne Mushayahama, Sternberg, P. W., Thomas, P., Van, K. M., & Ramsey, J. (2023). The Gene Ontology Knowledgebase in 2023. Genetics, 224(1). https://doi.org/10.1093/genetics/iyad031\n",
        "\n",
        "‌Ashburner, M., Ball, C. A., Blake, J. A., Botstein, D., Butler, H., Cherry, J. M., Davis, A. P., Dolinski, K., Dwight, S. S., Eppig, J. T., Harris, M. A., Hill, D. P., Issel-Tarver, L., Kasarskis, A., Lewis, S., Matese, J. C., Richardson, J. E., Ringwald, M., Rubin, G. M., & Sherlock, G. (2000). Gene Ontology: tool for the unification of biology. Nature Genetics, 25(1), 25–29. https://doi.org/10.1038/75556\n",
        "\n",
        "Chen, H., McCaffery, J. M., & Chan, D. C. (2007). Mitochondrial fusion protects against neurodegeneration in the cerebellum. Cell, 130(3), 548–562. https://doi.org/10.1016/j.cell.2007.06.026\n",
        "\n",
        "Fang, Z., Liu, X., & Peltz, G. (2022). GSEApy: a comprehensive package for performing gene set enrichment analysis in Python. Bioinformatics, 39(1). https://doi.org/10.1093/bioinformatics/btac757\n",
        "\n",
        "Feature Extraction — neuron morphology 1.1.7 documentation. (2023). Readthedocs.io. https://neuron-morphology.readthedocs.io/en/latest/feature_extraction.html\n",
        "\n",
        "Gouwens, N. W., Sorensen, S. A., Baftizadeh, F., Budzillo, A., Lee, B. R., Jarsky, T., Alfiler, L., Baker, K., Barkan, E., Berry, K., Bertagnolli, D., Bickley, K., Bomben, J., Braun, T., Brouner, K., Casper, T., Crichton, K., Daigle, T. L., Dalley, R., & de Frates, R. A. (2020). Integrated Morphoelectric and Transcriptomic Classification of Cortical GABAergic Cells. Cell, 183(4), 935-953.e19. https://doi.org/10.1016/j.cell.2020.09.057\n",
        "\n",
        "Hagenston, A. M., & Bading, H. (2011). Calcium Signaling in Synapse-to-Nucleus Communication. Cold Spring Harbor Perspectives in Biology, 3(11), a004564–a004564. https://doi.org/10.1101/cshperspect.a004564\n",
        "\n",
        "Ishihara, N., Nomura, M., Jofuku, A., Kato, H., Suzuki, S. O., Masuda, K., Otera, H., Nakanishi, Y., Nonaka, I., Goto, Y., Taguchi, N., Morinaga, H., Maeda, M., Takayanagi, R., Yokota, S., & Mihara, K. (2009). Mitochondrial fission factor Drp1 is essential for embryonic development and synapse formation in mice. Nature Cell Biology, 11(8), 958–966. https://doi.org/10.1038/ncb1907\n",
        "\n",
        "‌Kanehisa, M., & Goto, S. (2000). KEGG: Kyoto Encyclopedia of Genes and Genomes. Nucleic Acids Research, 28(1), 27–30. https://doi.org/10.1093/nar/28.1.27\n",
        "\n",
        "Kimura, T., & Murakami, F. (2014). Evidence That Dendritic Mitochondria Negatively Regulate Dendritic Branching in Pyramidal Neurons in the Neocortex. Journal of Neuroscience, 34(20), 6938–6951. https://doi.org/10.1523/jneurosci.5095-13.2014\n",
        "\n",
        "Konur, S., & Ghosh, A. (2005). Calcium Signaling and the Control of Dendritic Development. Neuron, 46(3), 401–405. https://doi.org/10.1016/j.neuron.2005.04.022\n",
        "\n",
        "Kulkarni, V. A., & Firestein, B. L. (2012). The dendritic tree and brain disorders. Molecular and Cellular Neuroscience, 50(1), 10–20. https://doi.org/10.1016/j.mcn.2012.03.005\n",
        "\n",
        "Li, Z., Okamoto, K.-I., Hayashi, Y., & Sheng, M. (2004). The Importance of Dendritic Mitochondria in the Morphogenesis and Plasticity of Spines and Synapses. Cell, 119(6), 873–887. https://doi.org/10.1016/j.cell.2004.11.003\n",
        "\n",
        "McAllister, A. K. (2000). Cellular and Molecular Mechanisms of Dendrite Growth. Cerebral Cortex, 10(10), 963–973. https://doi.org/10.1093/cercor/10.10.963\n",
        "\n",
        "M Orlic-Milacic, Beavers, D., Conley, P., Gong, C., Gillespie, M., Griss, J., Haw, R., Jassal, B., Matthews, L., May, B., Petryszak, R., Eliot Ragueneau, Rothfels, K., Sevilla, C., Shamovsky, V., Stephan, R., Krishna Kumar Tiwari, Thawfeek Varusai, Weiser, J., & Wright, A. (2023). The Reactome Pathway Knowledgebase 2024. Nucleic Acids Research, 52(D1). https://doi.org/10.1093/nar/gkad1025\n",
        "\n",
        "Redmond, L., Kashani, A. H., & Ghosh, A. (2002). Calcium Regulation of Dendritic Growth via CaM Kinase IV and CREB-Mediated Transcription. Neuron, 34(6), 999–1010. https://doi.org/10.1016/s0896-6273(02)00737-7\n",
        "\n",
        "Sokal, R., & Michener, C. (1958). A Statistical Method for Evaluating Systematic Relationships. University of Kansas Science Bulletin. https://ia800703.us.archive.org/5/items/cbarchive_33927_astatisticalmethodforevaluatin1902/astatisticalmethodforevaluatin1902.pdf\n",
        "\n",
        "Subramanian, A., Tamayo, P., Mootha, V. K., Mukherjee, S., Ebert, B. L., Gillette, M. A., Paulovich, A., Pomeroy, S. L., Golub, T. R., Lander, E. S., & Mesirov, J. P. (2005). Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles. Proceedings of the National Academy of Sciences, 102(43), 15545–15550. https://doi.org/10.1073/pnas.0506580102\n",
        "\n",
        "‌Wong, R. O. L., & Ghosh, A. (2002). Activity-dependent regulation of dendritic growth and patterning. Nature Reviews Neuroscience, 3(10), 803–812. https://doi.org/10.1038/nrn941‌\n",
        "\n",
        "Wu, G.-Y. ., Deisseroth, K., & Tsien, R. W. (2001). Activity-dependent CREB phosphorylation: Convergence of a fast, sensitive calmodulin kinase pathway and a slow, less sensitive mitogen-activated protein kinase pathway. Proceedings of the National Academy of Sciences, 98(5), 2808–2813. https://doi.org/10.1073/pnas.051634198\n",
        "\n",
        "‌Wu, T., Hu, E., Xu, S., Chen, M., Guo, P., Dai, Z., Feng, T., Zhou, L., Tang, W., Zhan, L., Fu, X., Liu, S., Bo, X., & Yu, G. (2021). clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation, 2(3), 100141. https://doi.org/10.1016/j.xinn.2021.100141\n"
      ],
      "metadata": {
        "id": "HorlUrFQhzM2"
      }
    }
  ]
}